pyop2/codegen/rep2loopy.py

```diff
@@ -0,0 +1,782 @@
+            for d in domains:
+                if d.get_dim_name(isl.dim_type.set, 0) == "layer":
+                    # layer = t1 - 1
+                    t1 = parameters.layer_end
+                    new_domains.append(d.add_constraint(isl.Constraint.eq_from_names(d.space, {"layer": 1, t1: -1, 1: 1})))
+                else:
+                    new_domains.append(d)
+        domains = new_domains
+
+    assumptions, = reduce(operator.and_,
+                          parameters.assumptions.values()).params().get_basic_sets()
+    options = loopy.Options(check_dep_resolution=True, ignore_boostable_into=True)
+
+    # sometimes masks are not used, but we still need to create the function arguments
+    for i, arg in enumerate(parameters.wrapper_arguments):
+        if parameters.kernel_data[i] is None:
+            arg = loopy.GlobalArg(arg.name, dtype=arg.dtype, shape=arg.shape)
+            parameters.kernel_data[i] = arg
+
+    if wrapper_name is None:
+        wrapper_name = "wrap_%s" % builder.kernel.name
+
+    wrapper = loopy.make_kernel(domains,
+                                statements,
+                                kernel_data=parameters.kernel_data,
+                                target=loopy.CTarget(),
+                                temporary_variables=parameters.temporaries,
+                                symbol_manglers=[symbol_mangler],
+                                options=options,
+                                assumptions=assumptions,
+                                lang_version=(2018, 2),
+                                name=wrapper_name)
+
+    # additional assumptions
+    wrapper = loopy.assume(wrapper, "start <= end")
+    wrapper = loopy.assume(wrapper, "start >= 0")
+    if builder.extruded:
+        wrapper = loopy.assume(wrapper, "{0} <= {1}".format(parameters.layer_start, parameters.layer_end))
+
+    # prioritize loops
+    for indices in context.index_ordering:
+        wrapper = loopy.prioritize_loops(wrapper, indices)
+
+    # register kernel
+    kernel = builder.kernel
+    headers = set(kernel._headers)
+    headers = headers | set(["#include <math.h>"])
+    preamble = "\n".join(sorted(headers))
+
+    from coffee.base import Node
+
+    if isinstance(kernel._code, loopy.LoopKernel):
+        knl = kernel._code
+        wrapper = loopy.register_callable_kernel(wrapper, knl)
+        from loopy.transform.callable import _match_caller_callee_argument_dimension_
+        wrapper = _match_caller_callee_argument_dimension_(wrapper, knl.name)
+        wrapper = loopy.inline_callable_kernel(wrapper, knl.name)
+    else:
+        # kernel is a string, add it to preamble
+        if isinstance(kernel._code, Node):
+            code = kernel._code.gencode()
+        else:
+            code = kernel._code
+        wrapper = loopy.register_function_id_to_in_knl_callable_mapper(
+            wrapper,
+            PyOP2KernelLookup(kernel.name, code, tuple(builder.argument_accesses)))
```

Comment:
Sometimes this "kernel" might have multiple C functions defined, e.g.:

```
void foo(...) {
   foo_sub(...);
}

void foo_sub(...) {
   ...
}
```

Where we advertise `foo` as the external kernel. I presume this all works in that case too?

Reply:
If they are all strings, it SHOULD be fine.